Implement hooks

vaibhavhrt opened the pull request:
- [x] closes #2

vaibhavhrt committed:
convert the class Dragger into a function using hooks

vaibhavhrt commented:
Only thing remaining to do is write proptypes, I will do that tomorrow. @nickmcmillan take a look at the code, if it's looks alright to you, you can merge it after I write the proptypes..

nickmcmillan commented:
Thanks for making a start on this @vaibhavhrt. Did you test it out to ensure all of your changes worked? To me it looks like there's still a lot that needs to be done. Maybe you forgot to commit the rest of your work? For instance there's still references to `this.onMove` and `this.inputType`...

vaibhavhrt commented:
Oops, yeah you are right. I will commit and push when I open my laptop.

On Fri, 24 May 2019, 5:46 pm Nick McMillan, <notifications@github.com>
wrote:

> Thanks for making a start on this @vaibhavhrt
> <https://github.com/vaibhavhrt>. Did you test it out to ensure all of
> your changes worked? To me it looks like there's still a lot that needs to
> be done. Maybe you forgot to commit the rest of your work? For instance
> there's still references to this.onMove and this.inputType...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nickmcmillan/react-physics-dragger/pull/3?email_source=notifications&email_token=AFMJMLC54OPPJOVSCXKRF5LPW7MDNA5CNFSM4HPJCHGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWFEBEA#issuecomment-495599760>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFMJMLEAKCC3OMAZCOH5MBDPW7MDNANCNFSM4HPJCHGA>
> .
>

vaibhavhrt committed:
remove all `this` that was accidently left in

vaibhavhrt committed:
Merge branch 'master' into hooks

vaibhavhrt committed:
fix an oopsie

vaibhavhrt committed:
fix few more oopsie

vaibhavhrt commented:
@nickmcmillan it's not working, I have set up a code sandbox:
https://codesandbox.io/s/reactphysicsdraggerhookstest-hlyh3?fontsize=14

There is some error but there is no error message so I can't figure it out, maybe you can take a look.
Btw, maybe you should provide a code sandbox link in readme on website for people to try it out easily.

vaibhavhrt commented:
I am trying to debug it and found a few things, will fix them and see if it works then.

vaibhavhrt committed:
convert outerEl and innerEl to ref

vaibhavhrt committed:
conver outer and innerWidth to ref

vaibhavhrt committed:
convert most vars to ref and it's partially working now

vaibhavhrt commented:
@nickmcmillan I have converted some variables to `ref` because `useState` was causing problems as it was not updating when expected. Take a look at codesandbox, example 1 is working, example 2 & 3 is still erroring. I will try to figure it out. Meanwhile you can take a look to find out what's wrong if you get some time, as you have a better idea how the code works.

vaibhavhrt committed:
convert settings to ref as well

vaibhavhrt commented:
@nickmcmillan I have fixed all the issues, and it's working as expected now. You can take a look at the codesandbox to double check it. I will add proptypes and then you can merge it.
I have converted almost everything to `ref` instead `useState`, I don't think it should be an issue.
I have made a lot of commits so you should do a squash and merge to keep your git history clean.

vaibhavhrt committed:
add proptypes

vaibhavhrt commented:
@nickmcmillan I have configured travis.yml too, tests will run on travis now.

vaibhavhrt commented on .travis.yml:
```diff
@@ -1,4 +1,11 @@
+    - npm install
+  script:
+    - npm test
+    # - npm run lint
```
Set up a lint script in package.json then uncomment this to run eslint checks.

vaibhavhrt reviewed (commented):
In package.json:

       "test": "cross-env CI=1 react-scripts test --env=jsdom",

@nickmcmillan is there some specific reason to setting CI=1. Travis will set the env var `CI=True` automatically so we don't need to do that manually.

nickmcmillan commented:
This is looking great. I see what you mean about using `useState` and the value not being updated when expected. `useRef` seems to solve this well. Nicely done.

I have noticed that this problem still exists. Within `onStart` when we use `setInputType(e.type === 'mousedown' ? 'mouse' : 'touch')` and then attempt to read from `inputType` immediately after. I've found that replacing this with `useRef` (like you've done previously) seems to solve the issue. Do you see this as a good solution?

nickmcmillan commented:
Just another thing @vaibhavhrt - would you mind keeping this pull request related to hooks only? If you'd like to add testing could you make that a separate PR? Thanks!

vaibhavhrt commented:
OK, I will delete some commits.I think we need `ref` when changing values
inside a function and trying to use the updated values. State doesn't seems
to work.
 When a function is created it kinda snapshots all `state` and uses it so
updating state inside fuction doesn't works as expected.


On Mon, May 27, 2019 at 3:54 PM Nick McMillan <notifications@github.com>
wrote:

> Just another thing @vaibhavhrt <https://github.com/vaibhavhrt> - would
> you mind keeping this pull request related to hooks only? If you'd like to
> add testing could you make that a separate PR? Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nickmcmillan/react-physics-dragger/pull/3?email_source=notifications&email_token=AFMJMLDGPJVHHCVD37IBOW3PXOZEVA5CNFSM4HPJCHGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWJN3FY#issuecomment-496164247>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFMJMLCSOCBTOAO2QWQSPHLPXOZEVANCNFSM4HPJCHGA>
> .
>

nickmcmillan committed:
🚧 convert inputType to useRef

nickmcmillan committed:
♻️ splitting state into separate useRefs

Plus minor tidying up and commenting.

nickmcmillan committed:
♻️ splitting state into separate useRefs

Plus minor tidying up and commenting.

nickmcmillan committed:
🐛 useState is required for updating isDragging cursor styling

nickmcmillan committed:
🐛 useState is required for updating isDragging cursor styling

nickmcmillan committed:
♻️ removing children isRequired prop

A user might have a valid case for not rendering children

nickmcmillan committed:
♻️ removing children isRequired prop

A user might have a valid case for not rendering children

vaibhavhrt commented:
I think everything is done now. I have deleted CI related commits, you can merge it.

nickmcmillan commented:
Yep looks good! However I think there's still some room for improvement. 

As we discussed earlier, we've used `useRef` so that we can read from the variable immediately (`useState` values were not updating when we needed it). The trouble with `useRef` however is that when a `useRef` value changes it does not trigger a render, and therefore the styles are not updated (mouse cursor changing from grab to grabbing).

I've created `isDraggingStyle` which uses `useState`, which does correctly cause a render when changed. This solution works fine. But it seems odd to me that we'd need both. Just curious what are your thoughts on this? Is there another approach we've overlooked?

vaibhavhrt commented:
React docs says use `useCallback` instead of `useRef` to render on value change. https://reactjs.org/docs/hooks-reference.html#useref

vaibhavhrt commented:
I guess we need the `isDraggingStyle` state, but instead of updating it manually when `isDragging` ref is updated, we can convert `isDragging` to a `useCallback` and update the state in callback, which is a little better.

nickmcmillan committed:
♻️ switching to useCallback for functions

nickmcmillan committed:
📄 comments

nickmcmillan committed:
🐛 downX as ref, not state

nickmcmillan committed:
🐛 missed this prop

nickmcmillan commented:
Good call on using `useCallback`. I've used it to wrap the existing functions and pass in the useState variables. And the result seems to work ok. For reasons unknown to me yet it seems buggy in Firefox/OSX. For instance, try dragging multiple sliders quickly on the example page. Seems like the isDragging state indiscriminately gets stuck.

vaibhavhrt commented:
I have a windows system so can't check. Travis related commits keeps coming back, I think you should have force pushed your commits. Anyway let me know when all done, I will rebase one last time.

nickmcmillan committed:
♻️ reverting useCallback solution

nickmcmillan committed:
📦 fix for scaffolding tool not supporting hooks

https://github.com/transitive-bullshit/create-react-library/issues/99

nickmcmillan commented:
I wasn't particularly satisfied with the useCallback approach that I tried due to the onRelease function not reliably being triggered, causing the dragger to sometimes get "stuck" to the mouse. So I've reverted to the last best working solution, which is where we had both isDragging and isDraggingStyle.

Happy to merge it in.

nickmcmillan commented:
Released on npm as v1.1.0.
Thanks so much for your work on this @vaibhavhrt !
